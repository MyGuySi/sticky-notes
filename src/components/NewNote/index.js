// @flow

import type { NoteColor } from '../../types';
import React, { Component, createRef } from 'react';
import StickyNote from '../StickyNote';
import Stack from '../Stack';
import Input from '../Input';
import KeyboardShortcut from '../KeyboardShortcut';
import Button from '../Button';

type Props = {
  text: string,
  color: NoteColor,
  onChangeText: (text: string) => void,
  onClickSave: () => void,
  onClickCancel: () => void
}

class NewNote extends Component<Props> {
  inputEl: { current: null | HTMLInputElement }

  constructor (props: Props) {
    super(props);
    this.inputEl = createRef();
  }

  componentDidMount () {
    if (this.inputEl.current) {
      this.inputEl.current.focus();
    }
  }

  onSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    this.props.onClickSave();
  }

  onReset = (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    this.props.onClickCancel();
  }

  render () {
    return (
      <StickyNote color={this.props.color}>
        <KeyboardShortcut value='Escape' handler={this.props.onClickCancel} />
        <form onSubmit={this.onSubmit} onReset={this.onReset} data-testid='new-form'>
          <Stack>
            <Input
              ref={this.inputEl}
              placeholder='Type your message...'
              value={this.props.text}
              onChange={e => this.props.onChangeText(e.target.value)}
              data-testid='text-input'
            />
            <Stack direction='row'>
              <Button
                isPrimary
                type="submit"
                disabled={this.props.text === ''}
                data-testid='save-button'
              >
                Save (Enter)
              </Button>
              <Button
                type="reset"
                data-testid='cancel-button'
              >
                Cancel (Esc)
              </Button>
            </Stack>
          </Stack>
        </form>
      </StickyNote>
    )
  }
}

export default NewNote;