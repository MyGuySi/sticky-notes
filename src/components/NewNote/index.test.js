// @flow

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import NewNote from './index';

it('renders correctly', () => {
  const component = renderer.create(
    <NewNote
      text='Test'
      color='cyan'
      onChangeText={() => {}}
      onClickSave={() => {}}
      onClickCancel={() => {}}
    />
  )
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('calls change handler', () => {
  const handler = jest.fn();
  render(
    <NewNote
      text='Test'
      color='cyan'
      onChangeText={handler}
      onClickSave={() => {}}
      onClickCancel={() => {}}
    />
  )
  fireEvent.change(screen.getByTestId('text-input'), { target: { value: 'test' }});
  expect(handler.mock.calls.length).toBe(1);
});

it('calls save handler', () => {
  const handler = jest.fn();
  render(
    <NewNote
      text='Test'
      color='cyan'
      onChangeText={() => {}}
      onClickSave={handler}
      onClickCancel={() => {}}
    />
  )
  fireEvent.click(screen.getByTestId('save-button'));
  expect(handler.mock.calls.length).toBe(1);
});

it('calls cancel handler', () => {
  const handler = jest.fn();
  render(
    <NewNote
      text='Test'
      color='cyan'
      onChangeText={() => {}}
      onClickSave={() => {}}
      onClickCancel={handler}
    />
  )
  // Using fireEvent.click on the resert button seems to have issues with
  // propagating the onReset event, so decided to dispatch directly on the form.
  const resetEvent = new Event('reset');
  screen.getByTestId('new-form').dispatchEvent(resetEvent);
  expect(handler.mock.calls.length).toBe(1);
});