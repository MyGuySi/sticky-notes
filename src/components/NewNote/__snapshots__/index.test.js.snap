// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="note cyan"
>
  <form
    data-testid="new-form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="stack"
    >
      <div
        className="wrapper"
      >
        <input
          className="input"
          data-testid="text-input"
          onChange={[Function]}
          placeholder="Type your message..."
          value="Test"
        />
      </div>
      <div
        className="stack row"
      >
        <button
          className="button primary"
          data-testid="save-button"
          disabled={false}
          type="submit"
        >
          Save (Enter)
        </button>
        <button
          className="button"
          data-testid="cancel-button"
          type="reset"
        >
          Cancel (Esc)
        </button>
      </div>
    </div>
  </form>
</div>
`;
